### DISCUSSION TITLE
让应用更流畅的请求策略库Taro版🎉

### DISCUSSION THREAD

**Discussion by JOU-amjs** [NervJS/taro#13406] [生态]
嗨，各位！
最近我们发布了请求策略库alova的taro版本，你可以了解了解。

<p align="center">
  <img width="200px" src="https://user-images.githubusercontent.com/29848971/202826407-92f17a6a-bc69-4ef8-be6f-001413ca8c90.png" />
</p>
<p align="center"><a href="https://alova.js.org/">文档</a> | <a href="https://alova.js.org/extension/alova-adapter-taro">Taro适配器文档</a></p>

## alova 是什么
轻量级的请求策略库，它针对不同请求场景分别提供了具有针对性的请求策略，来提升应用可用性、流畅性，降低服务端压力，让应用如智者一般具备卓越的策略思维。alova 核心模块提供了各类适配器接口、中间件机制来保证高扩展能力，从而实现更多的请求场景。

## 为什么创造 alova

数据请求一直是应用程序必不可少的重要部分，自从 XMLHttpRequest 诞生以来请求方案层出不穷，客户端的数据交互探索一直聚焦于以下两个方向：

1. 请求实现越来越简单，如`$.ajax`、`axios`、`react-query`，以及`fetch api`等请求工具，编码形式从回调函数、Promise，再到 usehook 不断过渡，编码难度越来越小。
2. 数据交互提示越来越友好，大致经历了以下几个阶段：
   1. 无状态（卡死状态）
   2. loading 等待状态、进度条展示
   3. 骨架屏展示、旧数据占位展示

在这两个方面的探索已经足够多了，只是，在越来越看重的用户体验的时代，我们更应该关注应用的流畅性。人们总是将数据交互的性能归咎于网络状态，但我们到处都能看到频繁的重复请求现象。
而 alova 就是从请求策略入手，致力于提高应用的流畅性，好的请求策略不仅能提升性能和用户体验，还可以降低服务端压力。**alova 的使命，就是让应用更高效地管理 Client-Server 数据交互**，针对不同的请求场景提供更好的请求策略。
在以上的基础上，我们将请求场景进行抽象提出了 [请求场景模型（RSM）](https://alova.js.org/overview/RSM)，它很好地解释了 alova 的请求策略方案。

alova 具有很灵活的扩展能力来实现不同场景下的请求策略，我们对 alova 的预期是一个**简单编码即可实现特定场景的高效请求的请求策略工具**。

同时，在无感数据交互场景下，alova 已经走出了一大步，它让用户在一定程度上无需等待数据请求，以及提交即响应，**在接下来我们将会一个章节来讲解无感数据交互的相关内容**。

在未来，alova 将承载着我们对请求策略的探索之路继续前行。

> 目前支持`vue`、`react`、`svelte`，同时也支持`Uniapp`、`Taro`，更多框架支持敬请期待...

## 选择 alova 的理由

alova 可以让你在编写少量代码的情况下，提升应用在请求方面的流畅性，它提供了开箱即用的体验，而其他请求库更加关注网络请求的便利性。如果你希望不编写更多额外代码，也能获得流畅的应用，不妨来试试！

同时，alova 也提供了丰富的特性，满足 99%的请求需求，更完整的特性如下：

1. 🕶 支持 vue、react、svelte
2. 📑 与 axios 相似的 api 设计，更简单熟悉
3. 🍵 开箱即用的高性能请求策略，让应用更流畅
4. 🐦 4kb，只有 axios 的 30%+
5. 🔩 高灵活性，兼容任意请求库，如 axios、superagent 或 fetch-api
6. 🔋 3 种数据缓存模式，提升请求性能，同时降低服务端压力
7. 🔌 丰富的扩展功能，可以自定义请求适配器、存储适配器、中间件，以及 states hook
8. 💕 请求共享，避免同时发送相同请求
9. 🪑 数据预拉取，这意味着用户可以更快看到信息，无需等待
10. 🦾 实时自动状态管理
11. 🎪 交互式文档和示例
12. 🎈 Typescript 支持
13. ⚡ 支持 tree shaking，这意味着 alova 的生产体积往往小于 4kb

**Comment by tourze** [comment]
挺好的，swr外又有一个选择。

**Reply by JOU-amjs** [reply]
非常你的支持哦！这让我更有动力了😜

**Comment by ZakaryCode** [comment]
可以尝试给 Taro 文档提个 PR 哦

**Reply by JOU-amjs** [reply]
这这个建议不错，说干就干

**Comment by zhaihongbin** [comment]
请问支持rn吗

**Reply by JOU-amjs** [reply]
支持的哦，RN也有与web端相同的fetch

**Reply by zhaihongbin** [reply]
1. 配置是不是改为：
```js
const alovaInst = createAlova({
  baseURL: 'https://api.alovajs.org',
   ...AdapterTaro(),
});
```
2. rn报错：
Unable to resolve module alova/react from C:\workspace2\bimdos-taro\node_modules\@alova\adapter-taro\dist\alova-adapter-taro.umd.min.js

**Reply by JOU-amjs** [reply]
噢，你是在使用taro开发rn项目吧，是的，要改为上面那样。
是不是改成这样才报错的？是的话可以告诉我下你的alova和@alova/adapter-taro版本吗

**Reply by zhaihongbin** [reply]
只有rn报错，h5和weapp正常
"@alova/adapter-taro": "^1.1.1",
"alova": "^2.0.10",
"@tarojs/taro": "3.6.2",

**Reply by JOU-amjs** [reply]
明白了，你稍等，我来尝试一下

**Reply by JOU-amjs** [reply]
经过测试，原因是因为rn的metro版本到0.75.1以上才支持`unstable_enablePackageExports`，这个特性在alova有使用。https://facebook.github.io/metro/docs/configuration/#unstable_enablepackageexports-experimental

为了不影响其他库的版本，你可以在pakcage.json中这样写：
```json
"resolutions": {
    "**/metro-config": "0.76.0",
    "**/metro": "0.76.0"
  },
```
这样可以更改依赖库的依赖，加入后需要删除node_modules和lock文件，再使用`yarn install`来安装，亲测`npm install`会有问题，然后在metro.config.js中开启这个特性

```javascript
module.exports = mergeConfig({
  // custom your metro config here
  // https://facebook.github.io/metro/docs/configuration
  resolver: {
    unstable_enablePackageExports: true
  }
}, getMetroConfig())
```

**Reply by yoyo837** [reply]
npm 叫 overrides

**Reply by s0f** [reply]
用taro开发rn项目，如果使用useRequest，并且immediate没有设置为false，它不会默认发起请求。经过排查发现是变量isSSR为true导致的。
![image](https://github.com/NervJS/taro/assets/6438451/d108129f-65cc-4e2c-921d-666af4cbf337)
